File: solutions/374. Guess Number Higher or Lower.py
# The guess API is already defined for you.
# @param num, your guess
# @return -1 if my number is lower, 1 if my number is higher, otherwise return 0
# def guess(num: int) -> int:
​
class Solution:
    def guessNumber(self, n: int) -> int:
        #Stefan Pochmann, cleaner
        lo, hi = 1, n
        while lo < hi:
            mid = lo + ((hi-lo)>>1)
            if guess(mid) == 1:
                lo = mid + 1
            else:
                hi = mid
        return lo
        
        # #First attempt
        # l, r = 1, n
        # while l < r:
        #     choice = l + ((r-l)>>1)
        #     val = guess(choice)
        #     if val == 0:
        #         return choice
        #     elif val == 1:
        #         l = choice+1
        #     else:
        #         r = choice-1        
        # return l

File: solutions/374. Guess Number Higher or Lower.cpp
/** 
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return       -1 if num is higher than the picked number
 *                1 if num is lower than the picked number
 *               otherwise return 0
 * int guess(int num);
 */
​
class Solution {
public:
//     better
    int guessNumber(int n) {
        int l = 1, r = n;
        while (l < r) {
            int mid = l + (r-l)/2;
            int guessAns = guess(mid);
            if (guessAns == 0) 
                return mid;
            else if (guessAns == 1)
                l = mid + 1;
            else
                r = mid - 1;
        }
        return l;
    }
//     bad
//     int guessNumber(int n) {
//         // initialization
//         int left = 1, right = n, mid=left+(right-left)/2;
//         int guessAns = guess(mid);
//         while (guessAns != 0) {
//             if (guessAns == 1) left = mid+1;
//             else if (guessAns == -1) right = mid;
//             mid = left+(right-left)/2;
//             guessAns = guess(mid);
            
//         }
//         return mid;
//     }
};
